### PR TITLE
[auto-materialize] Add AutoMaterializeRule

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

Creates a (purely-internal, for now) AutoMaterializeRule class, as well as a RuleEvaluationContext class. AutoMaterializeRules contain functions which tell Dagster how to evaluate these rules.

An AutoMaterializePolicy can now be modeled as a collection of rules, but for now remains a set of booleans so as to not rock the boat on storage changes too much (as this will entail some backcompat struggles).

### CODE REVIEW

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    @staticmethod
+    def skip_on_parent_outdated() -> "SkipOnParentOutdatedRule":
+        return SkipOnParentOutdatedRule()
+
+    @abstractproperty
+    def decision_type(self) -> AutoMaterializeDecisionType:
+        ...
+
+    @abstractmethod
+    def evaluate(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:99]: This evaluates a single asset, right? If so, I think we should include `asset` in the name. Some docstring would alse be helpful.

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+        self, context: RuleEvaluationContext
+    ) -> Mapping[AutoMaterializeCondition, AbstractSet[AssetKeyPartitionKey]]:
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:101]: Having this return a `RuleEvaluationResult` would make it easier to add other stuff in the future and also make the signature a little shorter.

Also, if this is per-asset, using the `AssetKeyPartitionKey` type, which includes the asset keys seems redundant.

File: python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py [hunk 3]
```diff
@@ -132,6 +129,10 @@ def target_asset_keys_and_parents(self) -> AbstractSet[AssetKey]:
             for parent in self.asset_graph.get_parents(asset_key)
         } | self.target_asset_keys
 
+    @property
+    def respect_materialization_data_versions(self) -> bool:
+        return self._respect_materialization_data_versions
+
     def get_implicit_auto_materialize_policy(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py:136]: Not related to this PR, but can we get rid of this function now that the reconciliation sensor is removed?

File: python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py [hunk 4]
```diff
@@ -278,142 +282,6 @@ def get_asset_partitions_with_newly_updated_parents_for_key(
         ) = self._get_asset_partitions_with_newly_updated_parents_by_key_and_new_latest_storage_id()
         return updated_parent_mapping.get(asset_key, set())
 
-    def materializable_in_same_run(self, child_key: AssetKey, parent_key: AssetKey) -> bool:
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py:281]: <3 all the deletions in this file

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py [hunk 5]
```diff
@@ -120,17 +168,44 @@ def lazy(max_materializations_per_minute: Optional[int] = 1) -> "AutoMaterialize
...
+                AutoMaterializeRule.materialize_on_required_for_freshness(),
+                AutoMaterializeRule.skip_on_parent_outdated(),
+            },
             max_materializations_per_minute=check.opt_int_param(
                 max_materializations_per_minute, "max_materializations_per_minute"
             ),
         )
 
+    @public
+    def without_rules(self, *rules_to_remove: "AutoMaterializeRule") -> "AutoMaterializePolicy":
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py:184]: @schrockn do you generally believe in avoiding `*args` parameters like this? I think I recall you had a good reason for it, but I'm forgetting what it is.

**Review by sryza — COMMENTED** [review]
This looks like a big step in the right direction to me! I think there are a set of additional changes we'll want to make before making this interface more public, but I think we can merge this PR before making these changes:
- Move away from the name `AutoMaterializeCondition`, which I think is confusing with the new ontology.
- Narrow the `RuleEvaluationContext` interface. `CachingInstanceQueryer` has a lot on it, and we should see what the smallest subset is we can get to.

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py (hunk 5 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py:184]: Well seems like a good idea to default to not using it, as it limits your ability to add parameters in backwards compat way in the future

File: python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py (hunk 3 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py:136]: I believe @johannkm had a PR that did so, but it got reverted -- I forget why exactly, but yeah in principle there's no reason it needs to exist other than to keep the existing tests running as usual

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py (hunk 5 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py:184]: I think in this case, additional parameters could be added as keyword arguments, similar to https://sourcegraph.com/github.com/dagster-io/dagster/-/blob/python_modules/dagster/dagster/_core/definitions/asset_selection.py?L107. Won't protest too hard, just think it's a bit nicer to do:

```python
my_policy.with_rules(rule1, rule2)
```

vs

```python
my_policy.with_rules({rule1, rule2})
```

Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py:184]: I don't feel strongly

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py (hunk 2 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:101]: Definitely kinda redundant, but for the moment makes life a bit simpler, as it means unpartitioned and partitioned assets can share a return type. Would like to replace this in the near future, just keeping the surface area as small as possible for this PR

**Comment by OwenKephart** [comment]
> * Move away from the name `AutoMaterializeCondition`, which I think is confusing with the new ontology.

@sryza My thinking is that these can become `AutoMaterializeRuleEvaluation`s, giving us the following:

`AutoMaterializePolicy`: A set of `AutoMaterializeRule`s
`AutoMaterializeAssetEvaluation`: A set of `AutoMaterializeRuleEvaluation`s

and "When `AutoMaterializeRule`s are evaluated, they produce a mapping from `AutoMaterializeRuleEvaluation` to the set of partitions that they apply to".

**Review by sryza — APPROVED** [review]

**Pull request merged**